backend/app/business/open_food_facts/breeding_type_calculator.py

```diff
@@ -0,0 +1,223 @@
+
+    def __init__(self) -> None:
+        """
+        Initializes the repository by loading the breeding patterns from a predefined source.
+        """
+        self._patterns: dict[AnimalType, AnimalPatternMap] = self._load_patterns()
+
+    def get_patterns(self, animal_type: AnimalType) -> AnimalPatternMap:
+        """
+        Retrieve the patterns for a specific animal type.
+        Args:
+            animal_type (AnimalType)
+        Returns:
+            AnimalPatternMap: A dictionary containing both exact category tags and regex patterns
+            for the given animal type.
+        """
+        if animal_type not in self._patterns.keys():
+            return {("exact", "categories_tags"): {}, ("regex", "all"): {}}
+        return self._patterns.get(animal_type, {})
+
+    @staticmethod
+    def _load_patterns() -> dict[AnimalType, AnimalPatternMap]:
+        """
+        Loads the breeding patterns from a predefined set of data.
+        Returns:
+            dict[AnimalType, AnimalPatternMap]: A dictionary mapping animal types to breeding patterns.
+        """
+        return {
+            AnimalType.LAYING_HEN: {
+                ("exact", "categories_tags"): {
+                    LayingHenBreedingType.FREE_RANGE: {"en:free-range-chicken-eggs", "en:organic-eggs"},
+                    LayingHenBreedingType.BARN: {"en:barn-chicken-eggs"},
+                    LayingHenBreedingType.CAGE: {"en:cage-chicken-eggs"},
+                },
+                ("regex", "all"): {
+                    LayingHenBreedingType.FREE_RANGE: {get_free_range_regex()},
+                    LayingHenBreedingType.BARN: {get_barn_regex()},
+                    LayingHenBreedingType.CAGE: {get_cage_regex()},
+                },
+            },
+        }
+
+
+class BreedingTypeCalculator:
+    """
+    A calculator that determines the breeding type of a product based on its data.
+
+    This class analyzes a product data to determine the correct breeding type of
+    all animals types registered in BreedingPatternsRepository
+    """
+
+    def __init__(self, product_data: ProductData):
+        """
+        Initializes the calculator with the given product data.
+        Args :  product_data (ProductData)
+        """
+        self.product_data = product_data
+        self.patterns_repository = BreedingPatternsRepository()
+
+    def get_breeding_types_by_animal(self) -> dict[AnimalType, BreedingTypeAndWeight]:
+        """
+        Get the breeding types for all animal types based on the product data.
+        This method processes the product data for each animal type, determines if there is a
+        unique match for the breeding type, and returns the result.
+        Returns:
+            dict[AnimalType, BreedingTypeAndWeight]: A dictionary mapping each animal type to its respective
+            breeding type and weight = 0.
+        """
+        breeding_types_by_animal = {}
+
+        for animal_type in self.patterns_repository._patterns:
```

Comment:
Consider making the `animal_type` parameter optional in `get_patterns(self, animal_type: AnimalType = None)` and having it return all patterns when no specific animal type is provided. 

This would provide a cleaner public interface to access all supported animal types without directly accessing the private `_patterns` attribute. You can also make another getter if you prefer 👍 

```suggestion
        for animal_type in self.patterns_repository.get_patterns():
```

Reply:
Ok, I created a get_all_patterns method instead because I have a type issue with your solution

Thanks a lot :)